lib/ide/src/controller.rs

```diff
@@ -0,0 +1,426 @@
+        pub async fn create_module_controller
+        (&self, loc:&module::Location) -> FallibleResult<module::StrongHandle> {
+            let data = module::Data{
+                loc      : loc.clone(),
+                contents : self.read_module(loc.clone()).await?,
+                parent   : self.clone(),
+            };
+            let module = self.insert_module(data).await;
+            Ok(module)
+        }
+
+        /// Spawns the task that processes the file manager's events.
+        ///
+        /// Note that File Manager requires manually calling `process_events`
+        /// method to yield any events through its streams / futures.
+        /// (without such call, this processor will just wait indefinitely)
+        pub async fn run_fm_events
+        (&self, spawner:impl futures::task::LocalSpawn) -> FallibleResult<()> {
+            let fm_stream     = self.call_fm(|fm| fm.events()).await;
+            let processor_fut = fm_stream.for_each(move |_event| async {
+                // TODO [mwu] dispatch the notification to the appropriate module
+                todo!()
+            });
+
+            Ok(spawner.spawn_local(processor_fut)?)
+        }
+
+        /// Returns a module controller for given module location.
+        ///
+        /// Reuses existing controller if possible.
+        /// Creates a new controller if needed.
+        pub async fn open_module(&self, loc:&module::Location) -> FallibleResult<module::StrongHandle> {
+            if let Some(existing_controller) = self.lookup_module(loc.clone()).await {
+                Ok(existing_controller)
+            } else {
+                Ok(self.create_module_controller(&loc).await?)
+            }
+        }
+
+        /// Retuns a text controller for given file path. It may designate
+        /// either the Luna source file or other file belonging to the project.
+        ///
+        /// File should be an existing, correct UTF-8 encoded file.
+        pub async fn open_file_text(&self, _path:&std::path::Path) -> FallibleResult<text::StrongHandle> {
+            // TODO [mwu] similar to the above (will need to add a second map
+            //      to keep handles or perhaps abstract the handles map to a
+            //      reusable structure?
+            //      Also might need to prepare a module first.
+            todo!()
+        }
+    }
+
+    /// Boilerplate wrappers over stateful `Data` APIs. Refer to the `Data`
+    /// methods for documentation.
+    impl StrongHandle {
+        #[allow(missing_docs)]
+        pub async fn lookup_module
+        (&self, loc:module::Location)
+         -> Option<module::StrongHandle> {
+            self.with(move |data| data.lookup_module(&loc)).await
+        }
+        #[allow(missing_docs)]
+        pub async fn insert_module
+        (&self, module:module::Data)
+         -> module::StrongHandle {
+            self.with(|data| data.insert_module(module)).await
+        }
+        #[allow(missing_docs)]
```

Comment:
Why this has allow missing docs? Byt he name `call_fm` I completely do not know what this function does.

Reply:
this was refactored on my branch `wip/ao/text-controller`